### PR TITLE
Retrieve snapshot id from temporal date version

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Add support for dealing with `timestamp` and `date` temporal arguments in the `FOR TIMESTAMP AS OF` expressions used in the time-travel queries within the Iceberg connector.

### Testing
Use `register_table` and  `HiveMinioDataLake` to register a table containing content created in the past (e.g. : change local system date to 2022)

```
create table iceberg.default.nation with (format='parquet',location='s3://test-bucket/nation') as select * from tpch.sf1.nation
```
Persist the content of the table to `resources` directory.

**NOTE** Iceberg specifies the _complete path_ for the files referenced in the metadata so we need to register the table again exactly in the same location as the one from the content of the table.

## Non-technical explanation

Allow usage of `timestamp` or `date` arguments in the `FOR TIMESTAMP AS OF` expressions in time travel queries.
The session's timezone will be used to derive the actual timestamp with zone used in the query. 
See https://trino.io/docs/current/connector/iceberg.html#time-travel-queries

```
SELECT *
FROM iceberg.testdb.customer_orders FOR TIMESTAMP AS OF DATE '2022-03-23'
```

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Allow usage of `date` and `timestamp` argument in the `FOR TIMESTAMP AS OF` expressions in time travel queries.
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -4830,40 +4830,59 @@ private Optional<TableVersion> extractTableVersion(Table table, Optional<Express
...
+                    objectType instanceof TimestampType ||
+                    objectType instanceof DateType)) {
+                throw semanticException(TYPE_MISMATCH, queryPeriod, format(
+                        "Type %s invalid. Temporal pointers must be of type Timestamp, Timestamp with Time Zone, or Date.",
+                        objectType.getDisplayName()));
+            }
+            if (pointer == null) {
+                throw semanticException(INVALID_ARGUMENTS, queryPeriod, "Pointer value cannot be NULL");
+            }
+            Type targetType = createTimestampWithTimeZoneType(TimestampWithTimeZoneType.MAX_PRECISION);
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4865]: Do we want to allow the connector to still accept Dates, if it chooses so?

Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet)

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4865]: BTW with this change, `ConnectorTableVersion` should be explicit how the temporals are carried

Comment (findinpath) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4865]: > Do we want to allow the connector to still accept Dates, if it chooses so?

I'm having a hard time to figure out why would the connector prefer something else than timestamp with time zone.

> Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet) ?

Do you imply adding a method to `ConnectorMetadata` where the connector can state the maximum precision it expects? This sounds reasonable (at the cost of an extra operation on the `ConnectorMetadata` interface).

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:4865]: > > Do we want to allow the connector to still accept Dates, if it chooses so?
> 
> I'm having a hard time to figure out why would the connector prefer something else than timestamp with time zone.

not unreasonable

cc @martint 

> > Do we want to allow the connector to determine the precision of the timestamptz it wants? (i guess no-one wants picos yet) ?
> 
> Do you imply adding a method to `ConnectorMetadata` where the connector can state the maximum precision it expects? This sounds reasonable (at the cost of an extra operation on the `ConnectorMetadata` interface).

Depends on the answer to the first question (about data types / data type families)

**Comment by findepi** [comment]
@martint please take a look at the SPI part

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 2]
```diff
@@ -446,14 +448,22 @@ private static long getTargetSnapshotIdFromVersion(Table table, ConnectorTableVe
...
     {
         if (versionType instanceof TimestampWithTimeZoneType timeZonedVersionType) {
             long epochMillis = timeZonedVersionType.isShort()
                     ? unpackMillisUtc((long) version.getVersion())
                     : ((LongTimestampWithTimeZone) version.getVersion()).getEpochMillis();
             return getSnapshotIdAsOfTime(table, epochMillis);
         }
+        else if (versionType instanceof DateType) {
+            long epochMillis = LocalDate.ofEpochDay((Long) version.getVersion())
+                    .atStartOfDay()
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:461]: When the snapshot time is 2023-07-26 01:02:03 and SELECT query specifies `FOR TIMESTAMP AS OF DATE '2023-07-26`, the snapshot wouldn't be included, right?

Does this behavior conform to SQL standard? Could you leave the code comment?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+import static java.lang.String.format;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestIcebergReadVersionedTableByDate
+        extends AbstractTestQueryFramework
+{
+    private static final String BUCKET_NAME = "test-bucket";
+
+    private Minio minio;
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:43]: The metadata files use an absolute location to reference other files
e.g.

`"metadata-file" : "s3://test-bucket/nation/metadata/00000-7b71c3a2-077a-4885-bed4-d69907cad55a.metadata.json"`

I don't know how this can be done consistenly on the local fs.

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 4]
```diff
@@ -446,9 +448,18 @@ private static long getTargetSnapshotIdFromVersion(Table table, ConnectorTableVe
...
     {
-        if (versionType instanceof TimestampWithTimeZoneType timeZonedVersionType) {
+        if (versionType.equals(DATE)) {
+            // Retrieve the latest snapshot made before or at the beginning of the day of the specified date in the session's time zone
+            long epochMillis = LocalDate.ofEpochDay((Long) version.getVersion())
+                    .atStartOfDay()
+                    .atZone(session.getTimeZoneKey().getZoneId())
+                    .toInstant()
+                    .toEpochMilli();
+            return getSnapshotIdAsOfTime(table, epochMillis);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:460]: Redundant else.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+        minio.createBucket(BUCKET_NAME);
+
+        QueryRunner queryRunner = IcebergQueryRunner.builder()
+                .setIcebergProperties(
+                        ImmutableMap.<String, String>builder()
+                                .put("hive.s3.aws-access-key", MINIO_ACCESS_KEY)
+                                .put("hive.s3.aws-secret-key", MINIO_SECRET_KEY)
+                                .put("hive.s3.endpoint", minio.getMinioAddress())
+                                .put("hive.s3.path-style-access", "true")
+                                .put("hive.s3.streaming.part-size", "5MB") // minimize memory usage
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:62]: This property is for uploading files. I think we can remove this.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                                .put("hive.s3.max-connections", "8") // verify no leaks
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:63]: This property looks unrelated to the test. I would remove it.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        }
+    }
+
+    @Test
+    public void testSelectTableWithEndVersionAsDate()
+            throws Exception
+    {
+        String tableName = "test_iceberg_read_versioned_table_" + randomNameSuffix();
+        Path tableLocation = Files.createTempFile(tableName, null);
+        copyDirectoryContents(new File(Resources.getResource("iceberg/timetravel").toURI()).toPath(), tableLocation);
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:88]: These two lines look redundant.

File: plugin/trino-iceberg/src/test/resources/iceberg/timetravel/metadata/00000-e57ff5b7-992d-4030-a592-0cac0f224830.metadata.json [hunk 8]
```diff
@@ -0,0 +1,64 @@
+{
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/resources/iceberg/timetravel/metadata/00000-e57ff5b7-992d-4030-a592-0cac0f224830.metadata.json:1]: Can we add a README file having SQL statements?

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 9]
```diff
@@ -446,9 +448,18 @@ private static long getTargetSnapshotIdFromVersion(Table table, ConnectorTableVe
         return snapshotId;
     }
 
-    private static long getTemporalSnapshotIdFromVersion(Table table, ConnectorTableVersion version, io.trino.spi.type.Type versionType)
+    private static long getTemporalSnapshotIdFromVersion(ConnectorSession session, Table table, ConnectorTableVersion version, io.trino.spi.type.Type versionType)
     {
-        if (versionType instanceof TimestampWithTimeZoneType timeZonedVersionType) {
+        if (versionType.equals(DATE)) {
+            // Retrieve the latest snapshot made before or at the beginning of the day of the specified date in the session's time zone
```
[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:454]: Spark doesn't support time travel queries for date temporal values:
 `SELECT * FROM prod.db.table DATE AS OF '1986-01-21'`

Are you refferring to this method 

https://github.com/apache/iceberg/blob/91161185ce53abbaaee992ebc1d412052e87852b/core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java#L330-L345 

?


I don't understand why specifying this in this context would be relevant.

Do we eventually need time travel compatibility tests with Spark (follow-up) ?

[the remark this answers is not in the document]
Comment (findinpath) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:454]: Follow-up https://github.com/trinodb/trino/issues/18449

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java [hunk 10]
```diff
@@ -446,8 +448,17 @@ private static long getTargetSnapshotIdFromVersion(Table table, ConnectorTableVe
         return snapshotId;
     }
 
-    private static long getTemporalSnapshotIdFromVersion(Table table, ConnectorTableVersion version, io.trino.spi.type.Type versionType)
+    private static long getTemporalSnapshotIdFromVersion(ConnectorSession session, Table table, ConnectorTableVersion version, io.trino.spi.type.Type versionType)
     {
+        if (versionType.equals(DATE)) {
```
Comment (findepi) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java:453]: if we have DATE support, we should also have TIMESTAMP support too

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import static io.trino.testing.TestingNames.randomNameSuffix;
+import static io.trino.testing.containers.Minio.MINIO_ACCESS_KEY;
+import static io.trino.testing.containers.Minio.MINIO_SECRET_KEY;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestIcebergReadVersionedTableByDate
+        extends AbstractTestQueryFramework
+{
+    private static final String BUCKET_NAME = "test-bucket-time-travel";
```
Comment (findepi) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:35]: randomize (see comment in minioclient.createbucket)

**Review by findepi — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java (hunk 11 above)
Comment (findinpath) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java:35]: can't be done - registering a table with randomized bucket name affects the (absolute) paths referenced in the metadata files.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        queryRunner.execute("CREATE SCHEMA IF NOT EXISTS " + ICEBERG_CATALOG + ".tpch");
+        return queryRunner;
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void destroy()
+            throws Exception
+    {
+        if (minio != null) {
+            minio.close();
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByDate.java]: @findinpath I suppose the above suggestion means null out without `if`:
```java
    @AfterClass(alwaysRun = true)
    public void destroy()
    {
        minio = null; // closed by closeAfterClass
    }
```

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByTemporal.java [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+        assertThat(query(viennaSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF DATE '2023-07-03'"))
+                .matches("VALUES 1, 2");
+        assertThat(query(losAngelesSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF DATE '2023-07-03'"))
+                .matches("VALUES 1, 2, 3");
+        assertThat(query(viennaSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF DATE '2023-07-04'"))
+                .matches("VALUES 1, 2, 3");
+        assertThat(query(losAngelesSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF DATE '2023-07-04'"))
+                .matches("VALUES 1, 2, 3");
+
+        // version as timestamp
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/TestIcebergReadVersionedTableByTemporal.java:119]: Can we add more tests verifying time travel with milli seconds precision? e.g.
```java
        assertThat(query(utcSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF TIMESTAMP '2023-07-01 05:02:43.953'"))
                .returnsEmptyResult();
        assertThat(query(utcSession, "SELECT * FROM " + tableName + " FOR TIMESTAMP AS OF TIMESTAMP '2023-07-01 05:02:43.954'"))
                .matches("VALUES 1");
```

File: plugin/trino-iceberg/src/test/resources/iceberg/timetravel/README.md [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+In `iceberg.properties` add the following properties:
+
+```
+iceberg.unique-table-location=false
+iceberg.table-statistics-enabled=false
+```
+
+Use `trino` to create the table content
+
+```
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/resources/iceberg/timetravel/README.md:12]:
```suggestion
```sql
```

File: plugin/trino-iceberg/src/test/resources/iceberg/timetravel/README.md [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+CREATE TABLE iceberg.tiny.timetravel(data integer);
+-- increase the date on the host
+insert into iceberg.tiny.timetravel values 1;
+-- increase the date on the host
+insert into iceberg.tiny.timetravel values 2;
+-- increase the date on the host
+insert into iceberg.tiny.timetravel values 3;
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/resources/iceberg/timetravel/README.md:19]: nit: Uppercase "insert into" and "values".

**Comment by findinpath** [comment]
Rebased on `master` to address conflicts on the documentation (moving from .rst to .md).
No code changes.

File: docs/src/main/sphinx/connector/iceberg.md [hunk 16]
```diff
@@ -1329,6 +1329,24 @@ SELECT *
...
+
+```
+SELECT *
+FROM example.testdb.customer_orders FOR TIMESTAMP AS OF TIMESTAMP '2022-03-23 00:00:00'
+```
+
+```
+SELECT *
+FROM example.testdb.customer_orders FOR TIMESTAMP AS OF TIMESTAMP '2022-03-23 00:00:00.000 Europe/Vienna'
+```
```
Comment (ebyhr) [inline docs/src/main/sphinx/connector/iceberg.md:1348]: We need to scroll for reading the entire query. I think it would be better to wrap after the table name. Follow-up is fine because this format is same as the existing query. 

Screenshot 2023-09-06 at 17 39 00

**Comment by russch** [comment]
@findepi Thank you for great notes at the top of this PR :)
